### ISSUE TITLE
Human mouse movement typing

### ISSUE THREAD

**Issue by satanmp** [daijro/camoufox#19]
in Python, the mouse moves and inputs like a human, making the website unable to detect it,how can I do it, can you help me ,thanks?

**Comment by daijro** [comment]
Hello,

This is something I plan on adding to the Python library soon. 👍

**Comment by desoforgit** [comment]
> Hello,
> 
> This is something I plan on adding to the Python library soon. 👍
> 
> **Edit:** As mentioned in #7, theres a significant user base using Camoufox remotely, outside of the Camoufox library.
> 
> For that reason, I'll be working on implementing human-like mouse movement directly into Camoufox instead of the Python library.


Will it be optional?

**Comment by daijro** [comment]
> Will it be optional?

Yes, I will add a `humanize:cursor` property that can enable/disable it.

**Comment by satanmp** [comment]
> Hello,
> 
> This is something I plan on adding to the Python library soon. 👍
> 
> **Edit:** As mentioned in #7, theres a significant user base using Camoufox remotely, outside of the Camoufox library.
> 
> For that reason, I'll be working on implementing human-like mouse movement directly into Camoufox instead of the Python library.

thanks

**Comment by daijro** [comment]
Human-like mouse movement has been added in v130.0.1-beta.10.

**Comment by satanmp** [comment]
> Human-like mouse movement has been added in v130.0.1-beta.10.
> 
> Here is the demo from the README:
> 
>  humanize-demo.mp4

👍

**Comment by francadossantosjoaofelipe35** [comment]
how to enable

**Comment by daijro** [comment]
The `humanize` parameter has just been added to the pypi library.

I will keep this issue open for support for human-like typing.

**Comment by tobwen** [comment]
> Here is the demo from the README:

Hrhr, looks like it has the mode "drunk user" :-)

Thanks!

**Comment by tserebour** [comment]
tested the humanize click function. It works like a charm.
PS: it only works if you have click functionality in your playwright code.

**Comment by vlrevolution** [comment]
so the keyboard inputs are not yet humanized right? any way to achieve that? also how does mouse input work, it sens OS level input?

**Comment by tobwen** [comment]
> so the keyboard inputs are not yet humanized right? any way to achieve that

I wrote this a while ago, maybe it can help @daijro. But be warned: the randomizer might need a fix.
https://gist.github.com/tobwen/9c5d109e7bdfce163e4315a10c199f5d

**Comment by daijro** [comment]
This is the solution I have in hrequests. It's a bit rough though, and doesn't account for the distribution of keys very well.

https://github.com/daijro/hrequests/blob/v0.9.0/hrequests/browser/browser.py#L277

```python
    def type(self, selector: str, text: str, delay: int = 50, *, timeout: float = 30):
        for char in text:
            # Randomly enter characters with a 50% MOE
            self.page.keyboard.type(
                char, delay=randint(int(delay * 0.5), int(delay * 1.5))  # nosec
            )
```

**Comment by daijro** [comment]
> > so the keyboard inputs are not yet humanized right? any way to achieve that
> 
> I wrote this a while ago, maybe it can help @daijro. But be warned: the randomizer might need a fix. https://gist.github.com/tobwen/9c5d109e7bdfce163e4315a10c199f5d

I absolutely love this solution. May I implement it and credit you?

**Comment by JWriter20** [comment]
Closing — this is answered by the `humanize` option, which didn't exist in a documented form when the issue was filed.

```python
from camoufox.sync_api import Camoufox

with Camoufox(humanize=True) as browser:   # or humanize=1.5 to cap cursor travel time
    page = browser.new_page()
    ...
```

Mouse movement then follows generated trajectories rather than teleporting, and typing is dispatched with human-like intervals. Docs: https://camoufox.com/python/usage/

For ideas about extending it further — scroll behaviour, idle drift — **#259** is the active thread.
